Objects/listobject.c

```diff
@@ -2338,21 +2339,24 @@ list_sort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
             else {
                 ms.key_compare = safe_object_compare;
             }
+
+            if (keys_are_in_tuples) {
+                /* Make sure we're not dealing with tuples of tuples
+                * (remember: here, key_type refers list [key[0] for key in keys]) */
+                if (key_type == &PyTuple_Type) {
```

Comment:
```suggestion
                if ( (keys_are_all_same_type == 0) || (key_type == &PyTuple_Type) ) {
```

Reply:
And instead modify this check, which sets `tuple_elem_compare` to `safe_object_compare` in the case of nested tuples, to also fall back when the tuple first-elements are heterogeneous.

Reply:
I don't think that's going to work because of the early exit at https://github.com/python/cpython/pull/12209/files#diff-5037da2a492e5903f100107e0506f818R2310 you can't be sure that all keys are tuples anyway.

I think the simplest way forward would be to use two variables `key_type` and `inner_key_type` to know if it's `item[0]` or `item` that changed type.

Does this make sense?

Reply:
Yes, that sounds like the best way. Probably not good practice to have one variable doing two jobs anyway!